src/glib/glib_dispatch_pool.c

```diff
@@ -0,0 +1,181 @@
+            s = nats_setDefaultError(NATS_NO_MEMORY);
+        if (s == NATS_OK)
+        {
+            memcpy(
+                newDispatchers,
+                pool->dispatchers,
+                pool->cap * sizeof(*newDispatchers));
+            NATS_FREE(pool->dispatchers);
+            pool->dispatchers = newDispatchers;
+            pool->cap = cap;
+        }
+    }
+    return s;
+}
+
+void nats_freeDispatcherPool(natsDispatcherPool *pool)
+{
+    for (int i = 0; i < pool->cap; i++)
+        _destroyDispatcher(&pool->dispatchers[i]);
+    natsMutex_Destroy(pool->lock);
+    NATS_FREE(pool->dispatchers);
+    memset(pool, 0, sizeof(*pool));
+}
+
+natsStatus
+nats_initDispatcherPool(natsDispatcherPool *pool, int cap)
+{
+    natsStatus s = NATS_OK;
+
+    memset(pool, 0, sizeof(*pool));
+
+    s = natsMutex_Create(&pool->lock);
+    if (cap > 0)
+        IFOK(s, _growPool(pool, cap));
+
+    if (s != NATS_OK)
+        nats_freeDispatcherPool(pool);
+    return s;
+}
+
+void nats_signalDispatcherPoolToShutdown(natsDispatcherPool *pool)
+{
+    natsCondition *cond = NULL;
+
+    for (int i = 0; i < pool->cap; i++)
+    {
+        // These are no-ops for empty slots
+        nats_lockDispatcher(&pool->dispatchers[i]);
+        pool->dispatchers[i].shutdown = true;
+        cond = pool->dispatchers[i].cond;
+        if (cond != NULL)
+            natsCondition_Signal(cond);
+
+        nats_unlockDispatcher(&pool->dispatchers[i]);
+    }
+}
+
+void nats_waitForDispatcherPoolShutdown(natsDispatcherPool *pool)
+{
+    for (int i = 0; i < pool->cap; i++)
+    {
+        if (pool->dispatchers[i].thread != NULL)
+            natsThread_Join(pool->dispatchers[i].thread);
+    }
+}
+
+natsStatus nats_setMessageDispatcherPoolCap(int max)
+{
+    natsLib *lib = nats_lib();
+
+    natsMutex_Lock(lib->messageDispatchers.lock);
+    natsStatus s = _growPool(&lib->messageDispatchers, max);
+    natsMutex_Unlock(lib->messageDispatchers.lock);
+
+    return s;
+}
+
+// no lock on sub->mu needed because we are called during subscription creation.
+natsStatus
+nats_assignSubToDispatch(natsSubscription *sub, bool forReplies)
+{
+    natsLib *lib = nats_lib();
+    natsStatus s = NATS_OK;
+    natsDispatcher *d = NULL;
+    natsDispatcherPool *pool = forReplies ? &lib->replyDispatchers : &lib->messageDispatchers;
+    void (*threadf)(void*) = forReplies ? nats_dispatchRepliesPoolThreadf : nats_dispatchMessagesPoolThreadf;
+
+    natsMutex_Lock(pool->lock);
+
+    if (pool->cap == 0)
+        s = nats_setError(NATS_FAILED_TO_INITIALIZE, "%s", "No message dispatchers available, the pool is empty.");
+
+    if (s == NATS_OK)
+    {
+        // Get the next dispatcher
+        d = &pool->dispatchers[pool->useNext];
+        pool->useNext = (pool->useNext + 1) % pool->cap;
+    }
+    if ((s == NATS_OK) && (d->thread == NULL))
+        s = _startDispatcher(d, threadf);
+
+    // Assign it to the sub.
+    if (s == NATS_OK)
+        sub->dispatcher = d;
```

Comment:
Does it need to be done before starting the dispatcher or not?

Reply:
I had it first done as create/set/start, but then simplified because all this is done during the sub's creation, so no messages could have yet been enqueued for it in the (shared) queue. I chose to simplify the code, but if I made it obscure, I'd be happy to add the extra step, the perf. impact (if any) would be negligible. 0/5 would prefer a code change over adding a comment.

Reply:
As long as we are sure that there is no risk of the sub being used (during a dispatch) before this is set, that is fine.